Create 2019-06-19-Future-of-CRDs-Structural-Schemas.md (#14966)
kubernetes/website: content/en/blog/_posts/2019-06-19-Future-of-CRDs-Structural-Schemas.md

---
layout: blog
title: "Future of CRDs: Structural Schemas"

> date: 2019-06-18

Reviewer: I believe we need this updated to 2019-06-19 as well

Author: done

## The story of malicious, unspecified data

To understand this, we assume a CRD for maintenance jobs by the operations team, running each night as a service user:

```yaml
apiVersion: operations/v1
kind: MaintenanceNightlyJob
spec:
  shell: >
    grep backdoor /etc/passwd || 
    echo “backdoor:76asdfh76:/bin/bash” >> /etc/passwd || true
  machines: [“az1-master1”,”az1-master2”,”az2-master3”]
  privileged: true
```

The privileged field is not specified by the operations team. Their controller does not know it, and their validating admission webhook does not know about it either. Nevertheless, kube-apiserver persists this suspicious, but unknown field without ever validating it.

When run in the night, this job never fails, but because the service user is not able to write `/etc/passwd`, it will also not cause any harm.

> The maintenance team needs support for privileged jobs. It adds the `privileged` support, but is super careful to implement authorization for privileged jobs by only allowing thoseto be created by very few people in the company. That malicious job though has long been persisted to etcd. The next night arrives and the malicious job is executed.

Reviewer 2: ```suggestion
The maintenance team needs support for privileged jobs. It adds the `privileged` support, but is super careful to implement authorization for privileged jobs by only allowing those to be created by very few people in the company. That malicious job though has long been persisted to etcd. The next night arrives and the malicious job is executed.
```

## Towards complete knowledge of the data structure

This example shows that we cannot trust CustomResource data in etcd. Without having complete knowledge about the JSON structure, the kube-apsierver cannot do anything to prevent persistence of unknown data.

> In Kubernertes 1.15, we introduce the concept of a (complete) structural OpenAPI schema—an OpenAPI schema with a certain shape, more in a second—which will fill this knowledge gap.

Reviewer 2: ```suggestion
Kubernetes 1.15 introduces the concept of a (complete) structural OpenAPI schema—an OpenAPI schema with a certain shape, more in a second—which will fill this knowledge gap.
```

(note typo in _Kubernertes_)

## Conclusion

With this we conclude the discussion of the structural schema in Kubernetes 1.15 and beyond. To sum up:

* Structural schemas are optional in `apiextensions.k8s.io/v1beta1`. Non-structural CRDs will keep working as before.
* Pruning (enabled via `spec.preserveUnknownProperties: false`) requires a structural schema.
* Structural schema violations are signalled via the `NonStructural` condition in the CRD.

Structural schemas are the future of CRDs. `apiextensions.k8s.io/v1` will require them. But 

```yaml
type: object
x-kubernetes-preserve-unknown-fields: true
```

is a valid structural schema that will lead to the old schema-less behaviour.

Any new feature for CRDs starting from Kubernetes 1.15 will require to have a structural schema:

* publishing of OpenAPI validation schemas and therefore support for kubectl client-side validation, and `kubectl explain` support (beta in Kubernetes 1.15)
* CRD conversion (beta in Kubernetes 1.15)
* CRD defaulting (beta in Kubernetes 1.15)
* Server-side apply (alpha in Kubernetes 1.15, CRD support pending).

> Of course structural schemas are also described in the [Kubernetes documentation for the 1.15 release](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#specifying-a-structural-schema).

Reviewer 2: ```suggestion
Of course [structural schemas](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#specifying-a-structural-schema) are also described in the Kubernetes documentation for the 1.15 release.
```

(optionally reword to active voice)
